### PR TITLE
feat(s3/local): Allow for dynamic templated file loading

### PR DESCRIPTION
**Pull request by Christos-Hadjinikolis**
## Overview

Imagine a world where you want to conditionally load partitioned data, as per the below example: 

Screenshot 2024-03-12 at 17 03 58

Here, you want to read all data under a path-prefix, in this case: `s3://some-bucket/input/*` but include a condition that it needs to do so only if it include `runner-{id}`. 

What you need there is: 
* Ability to resolve path against a templated parameter;
* Ability to parse date in the directory structure, according to a wildcard path;

This PR addresses both of these requirements, bringing in a new feature to do so. 
## Key Changes

* Updated `WithS3PathPrefix` mixin to support conditional data loading from S3; 
* Updated `WithLocalBatch` mixin to support conditional local data loading;
* Updated `pydantic` validations to extend valid keys for `local_batch` and `s3_path_prefix` resource definitions;

## Impact

* New feature: Ability to conditionally load partitioned data;

## Checklist

The following won't apply in all cases - mark `N/A` next to point if not needed.

- [x] Appropriate unit tests added/updated
- [x] Module, function, class docstrings updated
- [x] Comments added to PR where necessary
- [x] Interface documentation updated
- [x] Semantic commits used for this PR, so that a CHANGELOG can be generated from them (don't delete your local branch! when tagging a new release from master)

## Optional Below:

- [x] Screenshots of actual interactions with live resources. 
- [x] Performance stats/graphs.

## Release-Steps

- [x] Merge PR
- [x] Release new tag

### CODE REVIEW

File: dynamicio/config/pydantic/io_resources.py [hunk 1]
```diff
@@ -116,59 +117,108 @@ def set_parent(self, parent: IOBinding):  # noqa: F821
...
+            raise ValueError("path_prefix must end with '/'")
+
+        # Check for dynamic_file_path format
+        if dynamic_file_path:
+            allowed_extensions = [".parquet", ".h5", ".json", ".csv"]
+            if not any(dynamic_file_path.endswith(ext) for ext in allowed_extensions):
+                raise ValueError("`dynamic_file_path` must end with one of: .parquet, .h5, .json, .csv")
+
+        return values
+
```
Comment (Christos-Hadjinikolis) [inline dynamicio/config/pydantic/io_resources.py:161]: We allow `dynamic_file_path` now, as an alternative key to `s3_path_prefix`

File: dynamicio/config/pydantic/io_resources.py [hunk 2]
```diff
@@ -116,59 +117,108 @@ def set_parent(self, parent: IOBinding):  # noqa: F821
...
```
Comment (Christos-Hadjinikolis) [inline dynamicio/config/pydantic/io_resources.py:216]: We allow dynamic_file_path now, as an alternative key to s3_path_prefix

File: dynamicio/mixins/with_s3.py [hunk 3]
```diff
@@ -158,39 +159,66 @@ def _write_to_s3_path_prefix(self, df: pd.DataFrame):
...
-        full_path_prefix = utils.resolve_template(f"s3://{bucket}/{path_prefix}", self.options)
+
+        if dynamic_file_path:
+            full_path = utils.resolve_template(f"s3://{bucket}/{dynamic_file_path}", self.options)
+        else:
+            full_path = utils.resolve_template(f"s3://{bucket}/{path_prefix}", self.options)
 
         # The `no_disk_space` option should be used only when reading a subset of columns from S3
-        if self.options.pop("no_disk_space", False):
+        if self.options.pop("no_disk_space", False) and path_prefix:
```
Comment (VRGhost) [inline dynamicio/mixins/with_s3.py:216]: for future reference: mutating the dict (via `pop()`) in an `if` condition is pretty evil

File: dynamicio/mixins/with_s3.py [hunk 4]
```diff
@@ -200,18 +228,35 @@ def _read_from_s3_path_prefix(self) -> pd.DataFrame:
...
-                temp_dir,
-                "--acl",
-                "bucket-owner-full-control",
-                "--only-show-errors",
-                "--exact-timestamps",
-            )
-
-            dfs = []
+            if dynamic_file_path:
+                prefix, suffix = full_path.rsplit("/**/", 1)
```
Comment (VRGhost) [inline dynamicio/mixins/with_s3.py:232]: fyi - this will make dynamicio *nix-only, it won't run on Windows hosts.

**Review by VRGhost — APPROVED** [review]
Left couple comments, but it does look acceptable.

**Pull request merged**